File: trueforce/wheel.py
"""
trueforce.wheel
===============
High-level API for TrueForce haptic output on the Logitech RS50.
"""

import math
import time
from typing import Callable, Optional

import usb.core

from .protocol import (
    EP_OUT, EP_IN,
    UPDATE_HZ, PAIRS_PER_PKT, SHIFT_PER_PKT, SAMPLE_RATE,
    TF_GAIN_DEFAULT, NEUTRAL_U16,
    build_packet, SHUTDOWN_PKT,
)
from .device import open_device, close_device, send_init


class TrueForceWheel:
    """
    Context manager for TrueForce haptic output on the Logitech RS50.

    Usage
    -----
    with TrueForceWheel() as wheel:
        wheel.init()
        wheel.vibrate(freq=83.0, amp=0.1, duration=2.0)

    All amplitude values are in the range [0.0, 1.0] where:
      - 0.0  = silence
      - 0.005 = PCAP reference (real engine rumble in BeamNG)
      - 1.0  = full scale (±32767 counts)
    """

    def __init__(self, vendor_id: int = 0x046d, product_id: int = 0xc276):
        self._vendor_id  = vendor_id
        self._product_id = product_id
        self._dev        = None
        self._detached   = []
        self._reg        = 0
        self._sample_idx = PAIRS_PER_PKT - 1  # pre-fill buffer offset

    # ── Context manager ────────────────────────────────────────────

    def __enter__(self):
        self._dev, self._detached = open_device(self._vendor_id, self._product_id)
        return self

    def __exit__(self, *_):
        self.close()

    # ── Setup ──────────────────────────────────────────────────────

    def init(self, skip_setup: bool = True) -> None:
        """
        Send the CMD=05 init sequence.

        Parameters
        ----------
        skip_setup : Skip SETUP_CMDS that cause wheel rotation (default True).
                     Set False if you also need kinesthetic FFB.
        """
        send_init(self._dev, skip_setup=skip_setup)

    def close(self) -> None:
        """Silence output and reattach kernel drivers."""
        if self._dev is not None:
            close_device(self._dev, self._detached)
            self._dev = None

    # ── Core streaming ─────────────────────────────────────────────

    def stream(
        self,
        gen_fn:   Callable[[float], tuple[float, float, float]],
        duration: Optional[float] = None,
        gain:     int = TF_GAIN_DEFAULT,
        label:    str = "",
    ) -> None:
        """
        Stream TrueForce output at 1000 Hz.

        Parameters
        ----------
        gen_fn   : Callable(elapsed_seconds) -> (kinesthetic, tf_amp, tf_freq)
                   kinesthetic : float [-1.0, 1.0]  steering torque
                   tf_amp      : float [0.0, 1.0]   vibration amplitude
                   tf_freq     : float              vibration frequency in Hz
        duration : Seconds to run.  None = run until Ctrl+C.
        gain     : Protocol byte[11].  Keep at 0x0d.
        label    : Optional label printed to stdout.
        """
        interval = 1.0 / UPDATE_HZ
        t_start  = time.monotonic()
        t_next   = t_start
        sent = 0; fb = 0

        if label:
            print(f"\n{label}  (Ctrl+C to stop)\n")

        try:
            while True:
                now     = time.monotonic()
                elapsed = now - t_start
                if duration is not None and elapsed >= duration:
                    break
                if now < t_next:
                    time.sleep(t_next - now)
                t_next += interval

                kin, amp, freq = gen_fn(elapsed)
                pkt = build_packet(
                    self._reg, kin, amp, freq,
                    self._sample_idx, gain,
                )
                try:
                    self._dev.write(EP_OUT, pkt, timeout=10)
                    sent += 1
                except usb.core.USBError as e:
                    print(f"\nUSB error: {e}")
                    break
                try:
                    self._dev.read(EP_IN, 64, timeout=1)
                    fb += 1
                except usb.core.USBError:
                    pass

                self._reg = (self._reg + 1) & 0xFF
                self._sample_idx += SHIFT_PER_PKT

                if label and sent % 200 == 0:
                    print(
                        f"  {elapsed:6.2f}s  amp={amp:.4f}  freq={freq:.0f}Hz  fb={fb}",
                        end="\r",
                    )
        except KeyboardInterrupt:
            print("\nStopped.")

        if label:
            print(f"\nDone: {sent} packets, {fb} responses")

    def stream_waveform(
        self,
        waveform_fn: Callable[[float], float],
        amp:         float,
        duration:    Optional[float] = None,
        gain:        int = TF_GAIN_DEFAULT,
        label:       str = "",
    ) -> None:
        """
        Stream a custom waveform function.

        Parameters
        ----------
        waveform_fn : Callable(t: float) -> float [-1.0, 1.0]
                      Pure waveform shape; amplitude is applied separately.
        amp         : Overall amplitude [0.0, 1.0].
        duration    : Seconds to run.  None = run until Ctrl+C.

        Example
        -------
        # Sawtooth wave at 60 Hz
        wheel.stream_waveform(
            lambda t: (t * 60 % 1.0) * 2 - 1,
            amp=0.05,
            duration=3.0,
        )
        """
        interval = 1.0 / UPDATE_HZ
        t_start  = time.monotonic()
        t_next   = t_start
        sent = 0

        if label:
            print(f"\n{label}  (Ctrl+C to stop)\n")

        try:
            while True:
                now     = time.monotonic()
                elapsed = now - t_start
                if duration is not None and elapsed >= duration:
                    break
                if now < t_next:
                    time.sleep(t_next - now)
                t_next += interval

                pkt = build_packet(
                    self._reg, 0.0, amp, 0.0,
                    self._sample_idx, gain,
                    waveform_fn=waveform_fn,
                )
                try:
                    self._dev.write(EP_OUT, pkt, timeout=10)
                    sent += 1
                except usb.core.USBError as e:
                    print(f"\nUSB error: {e}")
                    break
                try:
                    self._dev.read(EP_IN, 64, timeout=1)
                except usb.core.USBError:
                    pass

                self._reg = (self._reg + 1) & 0xFF
                self._sample_idx += SHIFT_PER_PKT
        except KeyboardInterrupt:
            print("\nStopped.")

        if label:
            print(f"\nDone: {sent} packets")

    # ── High-level effects ─────────────────────────────────────────

    def vibrate(
        self,
        freq:     float,
        amp:      float,
        duration: Optional[float] = None,
        gain:     int = TF_GAIN_DEFAULT,
    ) -> None:
        """
        Continuous sine-wave vibration.

        Parameters
        ----------
        freq     : Frequency in Hz.  Engine range: 40–200 Hz.
        amp      : Amplitude [0.0, 1.0].  PCAP reference: ~0.005.
        duration : Seconds.  None = until Ctrl+C.
        """
        self.stream(
            lambda t: (0.0, amp, freq),
            duration=duration,
            gain=gain,
            label=f"vibrate  freq={freq:.0f}Hz  amp={amp:.4f}",
        )

    def bump(
        self,
        freq:     float = 150.0,
        decay:    float = 20.0,
        interval: float = 0.5,
        duration: float = 6.0,
    ) -> None:
        """
        Repeating decaying impulse — road bump / curb effect.

        Parameters
        ----------
        freq     : Carrier frequency in Hz.
        decay    : Exponential decay rate.  Higher = shorter bump.
        interval : Seconds between bumps.
        duration : Total playback duration.
        """
        def gen(t: float):
            phase = (t % interval) / interval
            amp   = math.exp(-phase * decay) if phase < 0.3 else 0.0
            return 0.0, amp, freq

        self.stream(gen, duration=duration,
                    label=f"bump  freq={freq:.0f}Hz  interval={interval}s")

    def freq_sweep(
        self,
        freq_start: float = 10.0,
        freq_end:   float = 500.0,
        amp:        float = 1.0,
        duration:   float = 10.0,
    ) -> None:
        """
        Linear frequency sweep from freq_start to freq_end.

        Parameters
        ----------
        freq_start : Start frequency in Hz.
        freq_end   : End frequency in Hz.
        amp        : Amplitude [0.0, 1.0].
        duration   : Sweep duration in seconds.
        """
        def gen(t: float):
            hz = freq_start + (freq_end - freq_start) * min(t / duration, 1.0)
            return 0.0, amp, hz

        self.stream(gen, duration=duration,
                    label=f"freq_sweep  {freq_start:.0f}→{freq_end:.0f}Hz  amp={amp:.2f}")

    def amp_ramp(
        self,
        freq:      float = 83.0,
        amp_start: float = 0.0,
        amp_end:   float = 1.0,
        duration:  float = 5.0,
    ) -> None:
        """
        Linearly ramp amplitude from amp_start to amp_end.

        Parameters
        ----------
        freq      : Carrier frequency in Hz.
        amp_start : Starting amplitude [0.0, 1.0].
        amp_end   : Ending amplitude [0.0, 1.0].
        duration  : Ramp duration in seconds.
        """
        def gen(t: float):
            amp = amp_start + (amp_end - amp_start) * min(t / duration, 1.0)
            return 0.0, amp, freq

        self.stream(gen, duration=duration,
                    label=f"amp_ramp  {amp_start:.3f}→{amp_end:.3f}  freq={freq:.0f}Hz")

    def engine(
        self,
        rpm_fn:   Callable[[float], float],
        duration: Optional[float] = None,
        rpm_min:  float = 800.0,
        rpm_max:  float = 8000.0,
        freq_min: float = 40.0,
        freq_max: float = 200.0,
        amp_min:  float = 0.002,
        amp_max:  float = 0.025,
    ) -> None:
        """
        Simulate engine rumble tied to an RPM value over time.

        Parameters
        ----------
        rpm_fn   : Callable(elapsed) -> RPM float.
        duration : Seconds.  None = until Ctrl+C.
        rpm_min/max   : RPM range to map from.
        freq_min/max  : Frequency range to map to.
        amp_min/max   : Amplitude range to map to.

        Example
        -------
        # Linearly ramp RPM from 800 to 6000 over 10 seconds
        wheel.engine(
            rpm_fn=lambda t: 800 + (6000 - 800) * min(t / 10, 1.0),
            duration=10.0,
        )
        """
        def gen(t: float):
            rpm  = max(rpm_min, min(rpm_max, rpm_fn(t)))
            norm = (rpm - rpm_min) / (rpm_max - rpm_min)
            freq = freq_min + norm * (freq_max - freq_min)
            amp  = amp_min  + norm * (amp_max  - amp_min)
            return 0.0, amp, freq

        self.stream(gen, duration=duration, label="engine rumble")

    def stop(self) -> None:
        """Send a single silence packet immediately."""
        pkt = build_packet(self._reg, 0.0, 0.0, 0.0, self._sample_idx)
        try:
            self._dev.write(EP_OUT, pkt, timeout=10)
        except usb.core.USBError:
            pass
        self._reg = (self._reg + 1) & 0xFF
        self._sample_idx += SHIFT_PER_PKT

    def kinesthetic(self, torque: float, duration: float) -> None:
        """
        Apply constant kinesthetic (steering) torque.

        Parameters
        ----------
        torque   : [-1.0, 1.0]  negative = left, positive = right.
        duration : Seconds.
        """
        self.stream(
            lambda t: (torque, 0.0, 0.0),
            duration=duration,
            label=f"kinesthetic  torque={torque:+.2f}",
        )

File: trueforce/wheel.cpp
#include "wheel.hpp"
#include <algorithm>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cmath>
#include <thread>
#include <tuple>

namespace trueforce {

using clk  = std::chrono::steady_clock;
using ns   = std::chrono::nanoseconds;
using fsec = std::chrono::duration<float>;

static const ns PACKET_INTERVAL = std::chrono::duration_cast<ns>(
    std::chrono::duration<double>{1.0 / UPDATE_HZ});

static volatile bool g_stop = false;
static void handle_sigint(int) { g_stop = true; }

TrueForceWheel::TrueForceWheel(uint16_t vid, uint16_t pid) {
    dev_ = open_device(vid, pid);
}
TrueForceWheel::~TrueForceWheel() { close(); }
void TrueForceWheel::init(bool skip_setup) { send_init(dev_, skip_setup); }
void TrueForceWheel::close() { dev_.close(); }

void TrueForceWheel::tick(const Packet& pkt) {
    send(dev_, pkt, 10);
    recv(dev_, 1);
    reg_        = (reg_ + 1) & 0xFF;
    sample_idx_ += SHIFT_PER_PKT;
}

void TrueForceWheel::stream(
    std::function<std::tuple<float,float,float>(float)> gen_fn,
    double      duration,
    uint8_t     gain,
    std::string label)
{
    if (!label.empty())
        std::printf("\n%s  (Ctrl+C to stop)\n\n", label.c_str());

    auto prev_sig = std::signal(SIGINT, handle_sigint);
    g_stop = false;

    auto t_start = clk::now();
    auto t_next  = t_start;
    int  sent = 0;

    while (true) {
        auto  now     = clk::now();
        float elapsed = std::chrono::duration_cast<fsec>(now - t_start).count();

        if (g_stop) { std::printf("\nStopped.\n"); break; }
        if (duration > 0.0 && elapsed >= static_cast<float>(duration)) break;

        if (now < t_next)
            std::this_thread::sleep_until(t_next);
        t_next += PACKET_INTERVAL;

        auto [kin, amp, freq] = gen_fn(elapsed);
        auto pkt = build_packet(reg_, kin, amp, freq, sample_idx_, gain);
        tick(pkt);
        ++sent;

        if (!label.empty() && sent % 200 == 0)
            std::printf("  %6.2fs  amp=%.4f  freq=%.0fHz\r",
                        (double)elapsed, (double)amp, (double)freq);
    }

    std::signal(SIGINT, prev_sig);
    if (!label.empty()) std::printf("\nDone: %d packets\n", sent);
}

void TrueForceWheel::stream_waveform(
    std::function<float(float)> waveform_fn,
    float       amp,
    double      duration,
    std::string label)
{
    if (!label.empty())
        std::printf("\n%s  (Ctrl+C to stop)\n\n", label.c_str());

    auto prev_sig = std::signal(SIGINT, handle_sigint);
    g_stop = false;

    auto t_start = clk::now();
    auto t_next  = t_start;
    int  sent = 0;

    while (true) {
        auto  now     = clk::now();
        float elapsed = std::chrono::duration_cast<fsec>(now - t_start).count();

        if (g_stop) { std::printf("\nStopped.\n"); break; }
        if (duration > 0.0 && elapsed >= static_cast<float>(duration)) break;

        if (now < t_next)
            std::this_thread::sleep_until(t_next);
        t_next += PACKET_INTERVAL;

        auto pkt = build_packet(reg_, 0.0f, amp, 0.0f, sample_idx_,
                                TF_GAIN_DEFAULT, waveform_fn);
        tick(pkt);
        ++sent;
    }

    std::signal(SIGINT, prev_sig);
    if (!label.empty()) std::printf("\nDone: %d packets\n", sent);
}

void TrueForceWheel::vibrate(float freq, float amp, double duration) {
    char label[64];
    std::snprintf(label, sizeof(label), "vibrate  freq=%.0f Hz  amp=%.4f", freq, amp);
    stream([freq, amp](float) { return std::make_tuple(0.0f, amp, freq); },
           duration, TF_GAIN_DEFAULT, label);
}

void TrueForceWheel::bump(float freq, float decay, float interval, double duration) {
    char label[64];
    std::snprintf(label, sizeof(label), "bump  freq=%.0f Hz  interval=%.2f s", freq, interval);
    stream([freq, decay, interval](float t) {
        float phase = std::fmod(t, interval) / interval;
        float amp   = (phase < 0.3f) ? std::exp(-phase * decay) : 0.0f;
        return std::make_tuple(0.0f, amp, freq);
    }, duration, TF_GAIN_DEFAULT, label);
}

void TrueForceWheel::freq_sweep(float freq_start, float freq_end, float amp, double duration) {
    char label[64];
    std::snprintf(label, sizeof(label), "freq_sweep  %.0f->%.0f Hz  amp=%.2f",
                  freq_start, freq_end, amp);
    float dur_f = static_cast<float>(duration);
    stream([freq_start, freq_end, amp, dur_f](float t) {
        float hz = freq_start + (freq_end - freq_start) * std::min(t / dur_f, 1.0f);
        return std::make_tuple(0.0f, amp, hz);
    }, duration, TF_GAIN_DEFAULT, label);
}

void TrueForceWheel::amp_ramp(float freq, float amp_start, float amp_end, double duration) {
    char label[64];
    std::snprintf(label, sizeof(label), "amp_ramp  %.3f->%.3f  freq=%.0f Hz",
                  amp_start, amp_end, freq);
    float dur_f = static_cast<float>(duration);
    stream([freq, amp_start, amp_end, dur_f](float t) {
        float amp = amp_start + (amp_end - amp_start) * std::min(t / dur_f, 1.0f);
        return std::make_tuple(0.0f, amp, freq);
    }, duration, TF_GAIN_DEFAULT, label);
}

void TrueForceWheel::engine(
    std::function<float(float)> rpm_fn,
    double duration,
    float rpm_min, float rpm_max,
    float freq_min, float freq_max,
    float amp_min,  float amp_max)
{
    stream([=](float t) {
        float rpm  = std::clamp(rpm_fn(t), rpm_min, rpm_max);
        float norm = (rpm - rpm_min) / (rpm_max - rpm_min);
        float freq = freq_min + norm * (freq_max - freq_min);
        float amp  = amp_min  + norm * (amp_max  - amp_min);
        return std::make_tuple(0.0f, amp, freq);
    }, duration, TF_GAIN_DEFAULT, "engine rumble");
}

void TrueForceWheel::stop() {
    auto pkt = build_packet(reg_, 0.0f, 0.0f, 0.0f, sample_idx_);
    tick(pkt);
}

void TrueForceWheel::kinesthetic(float torque, double duration) {
    char label[64];
    std::snprintf(label, sizeof(label), "kinesthetic  torque=%+.2f", torque);
    stream([torque](float) { return std::make_tuple(torque, 0.0f, 0.0f); },
           duration, TF_GAIN_DEFAULT, label);
}

} // namespace trueforce
